File: 1091_shortest_path_bin_matrix.py
# Nathan Zhu 12:46 am Just got back from point Lobos yesterday, done 28 leetcode problems today.  Not bad.
# Leetcode 1091 | medium | medium
# Category: BFS

# Nothing special here, but "before or right after you make a new pair to queue, you should set grid[x][y] right away."
# Kept timing out on a simple BFS soln before I did this.
def shortestPathBinaryMatrix(grid):
    """
    :type grid: List[List[int]]
    :rtype: int
    """
    N = len(grid)
    def get_neighbors(grid, row, col):
        ret = []
        for dr, dc in [(0, 1), (0, -1), (1, 1), (1, -1), (1, 0), (-1, 0), (-1, -1), (-1, 1)]:
            newr, newc = dr + row, dc + col
            if newr < len(grid) and newc < len(grid[0]) and newr >= 0 and newc >= 0 and grid[newr][newc] == 0:
                grid[newr][newc] = 1                     # MASSIVE SPEED IMPROVEMENTS PRUNING HERE. 
                ret.append((newr, newc))
        return ret
    
    if grid[0][0] == 1 or grid[N - 1][N - 1] == 1: return -1
    
    q = [(0, 0)]
    ret = 0
    while q:
        newq = []
        ret += 1
        for r, c in q:
            if r == N - 1 and c == N - 1: return ret
            # grid[r][c] = 1                              # JUST SETTING HERE LEADS TO TLE
            
            for n in get_neighbors(grid, r, c):
                newq.append(n)
        q = newq
        
    return -1
        
        

File: bfs/1091_shortest_path_bin_matrix.cpp
/* Nathan Zhu May 8th, 2020 STarting Salesforce a week-ush.
*  Leetcode 1091 | medium | damn, learned how to hash a vector today lmao
*  Category: BFS
*
*  Wrote my own hashing function lmao
*/


#include <vector>
#include <unordered_set>

using namespace std;


struct VHash{
    int operator()(const vector<int>& vec) const{
        int ret = 0;
        for(auto i : vec) ret ^= i;
        return ret;
    }
};


vector<vector<int>> get_neighbors(vector<vector<int>>& grid, int r, int c){
    vector<int> dr = {1, 0, -1, 0, 1, 1, -1, -1};
    vector<int> dc = {0, 1, 0, -1, 1, -1, -1, 1};
    vector<vector<int>> ret;
    
    for(int i = 0; i < dc. size(); ++i){
        int newr = dr[i] + r, newc = dc[i] + c;
        if(0 <= newr and newr < grid.size() and 0 <= newc and newc < grid[0].size() and grid[newr][newc] == 0){
            vector<int> pr = {newr, newc};
            grid[newr][newc] = 1;
            ret.push_back(pr);
        }
    }
    
    return ret;
    
}
    
int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
    unordered_set<vector<int>, VHash> visited;
    
    
    vector<vector<int>> q;
    if(grid[0][0] == 0) q.push_back({0, 0});
    int ret = 1;
    
    while(q.size()){
        vector<vector<int>> newq;
        
        for(auto& vec : q){
            int r = vec[0], c = vec[1];
            if (r == grid.size() - 1 and c == grid[0].size() - 1) return ret;
            for(auto& neigh : get_neighbors(grid, r, c)){
                if(visited.count(neigh)) continue;
                visited.insert(neigh);
                newq.push_back(neigh);
            }
            
        }
        ret += 1;
        q = vector<vector<int>>(newq.begin(), newq.end());
    }
    
    return -1;
    
    
}